### PR TITLE
Remove redundant code computing tap hashes

### PR DESCRIPTION
**Pull request by dr-orlovsky**

### CODE REVIEW

File: src/util/taproot.rs [hunk 1]
```diff
@@ -131,6 +131,21 @@ impl TapLeafHash {
...
+        let mut eng = TapBranchHash::engine();
+        if a < b {
+            eng.input(&a);
+            eng.input(&b);
+        } else {
+            eng.input(&b);
+            eng.input(&a);
+        };
+        TapBranchHash::from_engine(eng)
+    }
```
Comment (tcharding) [inline src/util/taproot.rs:146]: Perhaps:
```
        let (min, max) = if a < b { (a, b) } else { (b, a) };

        let mut eng = TapBranchHash::engine();

        eng.input(&min);
        eng.input(&max);

        TapBranchHash::from_engine(eng)
    }
```

File: src/util/taproot.rs [hunk 2]
```diff
@@ -561,16 +576,9 @@ impl NodeInfo {
...
-            eng.input(&a.hash);
-            eng.input(&b.hash);
-        } else {
-            eng.input(&b.hash);
-            eng.input(&a.hash);
-        };
+        let hash = TapBranchHash::from_node_hashes(a.hash, b.hash);
         Ok(Self {
-            hash: sha256::Hash::from_engine(eng),
+            hash: sha256::Hash::from_inner(hash.into_inner()),
```
Comment (tcharding) [inline src/util/taproot.rs:581]: Is this a 'normal' thing to do, convert between the various sha256 hash types? Should be be providing conversion functions over in `bitcoin_hashes`?

Comment (apoelstra) [reply src/util/taproot.rs:581]: We should probably provide a conveience method for "wrapped hashes" to their generic form (and vice-versa), yeah.

Sometimes hash conversions reflect perversions in Bitcoin, and these should appear perversely in our code :) but I don't think this is one of them.

In any case, for purposes of this PR the explicit conversion is fine.

File: src/util/taproot.rs [hunk 3]
```diff
@@ -131,6 +131,21 @@ impl TapLeafHash {
     }
 }
 
+impl TapBranchHash {
+    /// Computes branch hash given two hashes of the nodes underneath it.
+    pub fn from_node_hashes(a: sha256::Hash, b: sha256::Hash) -> TapBranchHash {
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply src/util/taproot.rs:136]: Yeah, I like having it be public. It seems like a generally useful thing to do.

File: src/util/taproot.rs (hunk 1 above)
Comment (apoelstra) [reply src/util/taproot.rs:146]: I like this, but I won't hold up the PR for it.

**Review by apoelstra — APPROVED** [review]
ACK f3ebfd6f8b049a2e2cfe6e9bd2b1b2f788b98d48

**Review by sanket1729 — APPROVED** [review]
sorta ACK 1b28375658d2016ea072a10b10fbebc362fa8514. Left one question, that I am unsure about

File: src/util/taproot.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dr-orlovsky) [reply src/util/taproot.rs:136]: There are three hash types involved in taproot trees: `TapBranchHash`, `TapLeafHash` and `sha256::Hash` for hidden branches/leaves or for parameters generic over branch/leaves. To solve the abiguity we have the following options
1. Add type alias `pub type TapNodeHash = sha256::Hash` and use it in API
2. Use `impl Into<TapNodeHash>` in arguments which are generic over specific hash type
3. Do `From<TapBranch|LeafHash> for TapNodeHash` convertors.

But even if we decide to do so, this should be a scope of separate (quite large) PR.

**Pull request merged**